src/main/java/org/codehaus/groovy/vmplugin/v9/Java9.java

```diff
@@ -81,15 +91,40 @@
         result.putAll(doFindClasses(URI.create("jrt:/modules/java.base/"), "java", javaPns));
 
         try {
-            URI gsLocation = DefaultGroovyMethods.getLocation(GroovySystem.class).toURI();
-            result.putAll(doFindClasses(gsLocation, "groovy", groovyPns));
+            List<URI> foundURIList = new LinkedList<>();
+            result.putAll(doFindGroovyClasses("groovy.lang.GroovySystem", groovyPns, foundURIList));
+
+            // in user production environment, Groovy core classes, e.g. `GroovySystem`(java class) and `GrapeIvy`(groovy class) are all packaged in the groovy core jar file,
+            // but in Groovy development environment, Groovy core classes are distributed in different directories
+            result.putAll(doFindGroovyClasses("groovy.grape.GrapeIvy", groovyPns, foundURIList));
         } catch (Exception e) {
-            System.err.println("[WARNING] Failed to get default imported groovy classes: " + e.getMessage());
+            throw new GroovyBugError("Failed to get default imported groovy classes", e);
         }
 
         return result;
     }
 
+    private Map<String, Set<String>> doFindGroovyClasses(String cn, List<String> groovyPns, List<URI> foundURIList) throws URISyntaxException {
+        GroovyClassLoader groovyClassLoader = new GroovyClassLoader();
+        URI classpathEntryURI;
+        try {
+            Class<?> clazz = groovyClassLoader.loadClass(cn);
+            classpathEntryURI = DefaultGroovyMethods.getLocation(clazz).toURI();
```

Comment:
I really wonder about this one. There is no gurantee that DefaultGroovyMethods.getLocation(clazz) will return something else than null, which would lead to a NPE here. In general I am wondering what happens if you are in an environment which does not provide the classes from file based locations. What if the classes actually come from a database? And yes, I mean the groovy-core classes itself. It is ok when it is slower then, as long as it still works. But the logic that handles the case it was not found in here it is not fully clear to me at this point actually

Reply:
> In general I am wondering what happens if you are in an environment which does not provide the classes from file based locations. 
I assume that groovy-core classes are all in some directories or a jar file. Even if groovy-core classes are stored in a database, they will have to written to file system at last.

To be frank, I have not found users store groovy-core classes in some non file system, e.g. database... In order to cover the edge cases, the PR provides a option to `groovy.deterministic.default.imports` to fallback.

Reply:
Even if they are stored locally, there is no guarantee that there is a location or that the location uses the normal protocols. I am thinking here application servers for example. All I am saying is that if the map cannot be populated, it should not make the compiler fail to resolve classes. Otherwise extensive tests on for example JBoss (for example with Groovy as module) are in order. We have no coverage in the test suite for things like that.

Reply:
> As 2 groovy elders worry about the impact of the PR, I close it now.
> Thanks for your revewing ;-)

I think the general idea is good, just I think we cannot rely on the cache in case of a miss. Could we forbid people putting classes in groovy.lang? In theory if we want to take the module system seriously then all the packages we use in core should be taboo for the users or else they would get the split package problem. But that also means to not allow people to fallback to the classpath environment / anonymous module Groovy actually needs atm. Such a decision should be done on the list.

Reply:
I agree with Jochen. More optimisation of imports is a good thing. Failing resolution for the case of a cache miss, rather than falling back to a slower path seems to me to be too big to do as a "tweak".

Reply:
Understood.
BTW, we have 2 "Proxy" default imports, one from JDK, another from GDK...

Reply:
That's unfortunate. The Java one takes precedence and I suspect the Groovy one is very rarely used. I am inclined to deprecate the Groovy one. We can copy it some place else or rename to GroovyProxy (or both), or we could just delete it. We would normally prefer @Delegate these days but I guess it is a dynamic equivalent and we even have better dynamic alternatives like ProxyGenerator, albeit with additional complexity. I'll create an issue.